tests/torchtune/models/mistral/test_mistral_classifier.py

```diff
@@ -0,0 +1,51 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import pytest
+import torch
+from tests.test_utils import fixed_init_model
+from torchtune.models.mistral import mistral_classifier
+from torchtune.utils.seed import set_seed
+
+EMBED_DIM = 128
+NUM_LAYERS = 4
+NUM_HEADS = 16
+NUM_KV_HEADS = 8
+VOCAB_SIZE = 32000
+MAX_SEQ_LEN = 2048
+BSZ = 2
+SEQ_LEN = 100
+NUM_CLASSES = 10
+INTERMEDIATE_DIM = 512
+
+
+@pytest.fixture(autouse=True)
+def random():
+    set_seed(16)
+
+
+class TestMistralClassifier:
+    @pytest.fixture
+    def inputs(self):
+        return torch.randint(0, VOCAB_SIZE, (BSZ, SEQ_LEN))
+
+    def test_forward(self, inputs):
+        model = mistral_classifier(
+            num_classes=NUM_CLASSES,
+            vocab_size=VOCAB_SIZE,
+            num_layers=NUM_LAYERS,
+            num_heads=NUM_HEADS,
+            num_kv_heads=NUM_KV_HEADS,
+            embed_dim=EMBED_DIM,
+            intermediate_dim=INTERMEDIATE_DIM,
+            max_seq_len=MAX_SEQ_LEN,
+        )
+        fixed_init_model(model, min_val=-0.25, max_val=0.5)
+        actual = model(inputs)
+        # TODO find the actual value
+        # expected = torch.tensor(3.9763)
+        assert actual.shape == (BSZ, SEQ_LEN, NUM_CLASSES)
+        # torch.testing.assert_close(actual.mean(), expected, atol=1e-4, rtol=1e-4)
```

Comment:
Need to update this prior to merging. Let me know if you need any assistance with the parity checks here

Reply:
I could use some guidance here :) Could you point to an existing test in the repo you think is a good example? I'm still working on the base `mistral` tests in another branch.

Reply:
(Note: this is a long comment, but for the actual practical suggestion for this particular PR just skip to the end)

The parity checks can be tricky, it also depends on the model and how readily available a solid reference implementation is. I will use our [test_lora.py](https://github.com/pytorch/torchtune/blob/main/tests/torchtune/modules/peft/test_lora.py) as a simple example here. We typically:

1) Find a reference implementation, and either copy-paste it or import it into a runnable Python script. We have a bunch of these under tests/torchtune/models/llama2/scripts (admittedly not the best location, but they are not really used and more meant as references). Here's the [copy-pasted LoRA layer](https://github.com/pytorch/torchtune/blob/main/tests/torchtune/models/llama2/scripts/compare_lora.py#L15-L111) from the original Microsoft repo.
2) In the same script, import our implementation. Then 
  a) set the random seed for reproducibility, create some dummy input tensor for the model, and initialize the weights of our implementation in a deterministic way. This is done [here](https://github.com/pytorch/torchtune/blob/main/tests/torchtune/models/llama2/scripts/compare_lora.py#L127-L141) using our test utility [fixed_init_model](https://github.com/pytorch/torchtune/blob/ea3d4eaca82d17fc83a979f81479c69c79dc9b92/tests/test_utils.py#L97)
  b) Load the weights we initialized for our model into the reference implementation. This usually involves remapping of state dicts (sometimes just a key mapping, sometimes some tensor reshaping). This part is usually fairly annoying, a bit of guess-and-check, and can actually get quite hairy for full models with nontrivial differences in how parameters are represented. [Example for the LoRA comparison](https://github.com/pytorch/torchtune/blob/main/tests/torchtune/models/llama2/scripts/compare_lora.py#L155-L162)
  c) Finally call forward on both models and check that the outputs line up. Then take some summary stat (like mean) to use as the expected value in the test. [LoRA example](https://github.com/pytorch/torchtune/blob/main/tests/torchtune/models/llama2/scripts/compare_lora.py#L166-L167)
3) Then in the unit test we use the value from (2) as our expected value ([in the LoRA test](https://github.com/pytorch/torchtune/blob/main/tests/torchtune/modules/peft/test_lora.py#L24)). We [set the same seed](https://github.com/pytorch/torchtune/blob/main/tests/torchtune/modules/peft/test_lora.py#L29) and do the [same initialization](https://github.com/pytorch/torchtune/blob/main/tests/torchtune/modules/peft/test_lora.py#L60) so that we get the same result from the comparison script.


If you've made it this far, you can see it's actually a fair amount of work to do things this way. We try to hit a really high bar on testing for correctness and have found this is the best way to do it. That being said, going through all these steps is definitely not a prerequisite for landing this change. We still need a proper comparison script for Mistral (though it was tested locally prior to landing) and shouldn't block your PR on that. 

Instead for testing this change, I would propose: assume the existing Mistral implementation in the repo is correct. Then the delta for your changes in terms of the Mistral builder is really just the change in output layer. Then the much-abbreviated version of the above is: (a) import an existing Mistral builder and hardcode the output projection to `nn.Linear(embed_dim, n_classes)`. Do the same initialization and run forward on that model to get the expected value. Then use this value in your unit test (no state dict remapping needed).

We can tackle the full testing for Mistral I outlined above in a separate PR, but definitely don't want to block this particular change on that. Let me know if this makes sense to you, and if you are still interested in adding the Mistral test following this approach in a separate PR I am happy to provide (even) more detailed guidance when needed.

Reply:
Thanks so much for this. I think some of it I grokked from the code but 2.b would've been tricky and sounds like it can be a pain in the ass. I've started this process with just the attention layer for mistral so far, I'll put a draft PR up to get your thoughts early on in this soon.

I think your suggestion for testing the classifier sounds sensible. It could be a pretty general test to make sure `output_layer` does what we want it to do.

Reply:
I've taken a go at writing a test following `test_lora.py`. Thoughts appreciated.